DYN-6094 Lucene unit testing

reddyashish opened the pull request:
### Purpose

This PR is to use Lucene search in the Dynamo unit tests. 

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB 

### Release Notes
DYN-6094 Lucene unit testing

### Reviewers
@QilongTang @RobertGlobant20 

### FYIs
@DynamoDS/dynamo

RobertGlobant20 commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -928,8 +927,16 @@ protected DynamoModel(IStartConfiguration config)
-            LuceneSearchUtility.InitializeLuceneConfig(LuceneConfig.NodesIndexingDirectory);
+            if (IsTestMode)
+            {
+                LuceneSearchUtility = new LuceneSearchUtility(this);
```
I think the  LuceneSearchUtility = new LuceneSearchUtility(this); line should be outside of the if sentence (line 929),

RobertGlobant20 commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1518,19 +1525,19 @@ private void InitializeIncludedNodes()
-            //}
-
-            var ele = AddNodeTypeToSearch(cbnData);
+            var ele = AddNodeTypeToSearch(outputData);      
```
just consider (for future bug fixes) that uncommenting this will probably generate duplicates in output and input nodes when searching in the Library (due that are already added in library).

reddyashish replied:
Thanks @RobertGlobant20. Will test this out.

RobertGlobant20 commented on src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:
```diff
@@ -937,7 +944,7 @@ internal IEnumerable<NodeSearchElementViewModel> Search(string search, IEnumerab
         /// <param name="useLucene"> Temporary flag that will be used for searching using Lucene.NET </param>
         internal IEnumerable<NodeSearchElementViewModel> Search(string search, bool useLucene)
         {
-            if (useLucene)
```
if `bool useLucene` is not used any more then the parameter in `Search(string search, bool useLucene)` should be also deleted, no?

reddyashish replied:
yes will update it.

RobertGlobant20 reviewed (approved):
LGTM with some comments

reddyashish commented on test/DynamoCoreWpfTests/WorkspaceSaving.cs:
```diff
@@ -1165,6 +1167,7 @@ public void CustomNodeSaveAsAddsNewCustomNodeToSearchAndItCanBeRefactoredWhilePr
         }
 
         [Test]
+        [Category("Failure")]
```
4 Tests are being marked as Failure for now as they are related to lucene indexing of Customnode renaming.

QilongTang reviewed (commented):
LGTM 
The regressions reported is because this branch is not up to date with master latest.
just curious did we have to update the result of any previous search tests?

RobertGlobant20 commented on src/DynamoCoreWpf/ViewModels/Search/NodeAutoCompleteSearchViewModel.cs:
```diff
@@ -614,10 +614,9 @@ private NodeSearchElementViewModel GetViewModelForNodeSearchElement(NodeSearchEl
+        internal IEnumerable<NodeSearchElementViewModel> SearchNodeAutocomplete(string search)
         {
-            if (useLucene)
+            if (LuceneSearchUtility != null)
```
here the validation is correct? or should using LuceneSearchUtilityNodeAutocomplete != null?

reddyashish replied:
yes, changing it.

RobertGlobant20 reviewed (approved):
LGTM with one comment

reddyashish commented:
@QilongTang updated to latest master now. did not have to change any search results for now.

QilongTang commented:
hi @reddyashish Let us know when this is ready for one more review

reddyashish commented on src/DynamoCore/Utilities/LuceneSearchUtility.cs:
```diff
@@ -67,28 +77,35 @@ internal void InitializeLuceneConfig(string dirName, LuceneStorage storageType =
+
+
+            // Create an analyzer to process the text
+            Analyzer = CreateAnalyzerByLanguage(dynamoModel.PreferenceSettings.Locale);
```
Added back the lucene search support for different languages.

RobertGlobant20 replied:
@reddyashish you miss to comment/remove the line 86 otherwise we will be using always the StandardAnalyzer.
![image](https://github.com/DynamoDS/Dynamo/assets/61755417/cf650fcb-d404-4ab1-8aae-18f608b2b5db)

reddyashish replied:
Addressing it in here: https://github.com/DynamoDS/Dynamo/pull/14248

reddyashish commented on src/DynamoCoreWpf/ViewModels/Search/SearchViewModel.cs:
```diff
@@ -937,55 +944,12 @@ internal IEnumerable<NodeSearchElementViewModel> Search(string search, IEnumerab
-            if (useLucene)
-            {
-                //The DirectoryReader and IndexSearcher have to be assigned after commiting indexing changes and before executing the Searcher.Search() method, otherwise new indexed info won't be reflected
-                LuceneSearchUtility.dirReader = LuceneSearchUtility.writer?.GetReader(applyAllDeletes: true);
```
All the search code is moved to the NodeSearchModel class.

reddyashish commented:
@QilongTang This is now ready for final review
Main changes in recent commit:
1) Updated SearchModelTests to use the lucene search model.
2) Moved the lucene search code from ViewModel to Model.
3) Added back the lucene search support for different languages.

reddyashish commented:
The failing tests here are from documentationbrowser and not related to this PR.

reddyashish commented:
Yes, I ran the self-serve and only 5 docbrowser tests are failing.

reddyashish commented:
Merging this and will cherrypick into 2.19
